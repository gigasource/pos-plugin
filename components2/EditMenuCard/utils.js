export function createEmptyProductLayout() {
  return {
    type: 'Article',
    text: '',
    color: '#ddd',
    product: createEmptyProduct()
  }
}

export function createEmptyProduct() {
  return {
    type: null,
    id: '',
    name: '',
    price: '',
    groupPrinter: null,
    groupPrinter2: null,
    isNoPrint: null,
    isItemNote: null,
    tax: null,
    tax2: null,
    category: null,
    isDivArticle: null
  }
}


export function createEmptyLayout(row, column) {
  return {
    top: row,
    left: column,
    name: ''
  }
}

export function createEmptyCategoryLayout() {
  return {
    rows: 10,
    columns: 6,
    color: '#FFF'
  }
}


export function isSameArea(area1, area2) {
  return area1.top === area2.top && area1.left === area2.left
}
