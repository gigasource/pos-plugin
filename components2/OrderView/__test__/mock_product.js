export const mockProduct = {
  "_id": "5e784d7786e3ec0997b5e39b",
  "category": {
    "_id": "5df6f5dd7854273fa4fd989e",
    "name": "Wine",
    "position": 5
  },
  "groupPrinter": {
    "_id": "5e7864b186e3ec0997b5e48e",
    "name": "Bar",
    "printer": "HP M176n",
    "printers": [
      {
        "_id": "5e81774685fb2d03dda95cd0",
        "printerType": "ip",
        "ip": "192.168.10.6",
        "escPOS": false,
        "merge": "true"
      }
    ],
    "showDineInTax": false,
    "type": "kitchen"
  },
  "id": "48",
  "layouts": [
    {
      "_id": "5e784d7786e3ec0997b5e39c",
      "color": "#DCE775",
      "order": 2,
      "device": "pos"
    }
  ],
  "name": "Whiskey",
  "option": {},
  "price": 40,
  "tax": 19,
  "tax2": 7
}
