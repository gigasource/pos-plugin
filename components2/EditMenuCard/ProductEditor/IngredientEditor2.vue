<script>
import _ from 'lodash';
import { Touch } from 'pos-vue-framework'
import { onActivated, ref, computed, watch, onMounted } from 'vue'
import { useI18n } from 'vue-i18n'
import { genScopeId } from '../../utils'
import {
  selectedProduct,
  selectedProductExisted,
  loadOrderLayout
} from '../../OrderView/pos-ui-shared'
import PosKeyboardFull from '../../pos-shared-components/PosKeyboardFull'
import {
  detailInventories
} from '../../Inventory/inventory-logic-ui'
import {loadInventories} from "../../Inventory/inventory-logic-be";
import {loadCategories} from "../../Product/product-logic-be";
import {loadProducts} from "../../Product/product-logic-be";

export default {
  name: 'IngredientEditor2',
  directives: {
    Touch
  },
  components: { PosKeyboardFull },
  props: {},
  setup() {
    const { t } = useI18n()
    const showKeyboard = ref(false)
    const ingredients = ref([])

    const productInventoryName = computed(() => {
      return detailInventories.value.map((inventory) => inventory.product.name)
    })

    function loadIngredients() {
      if (selectedProductExisted.value) {
        if (!selectedProduct.value.ingredients)
          selectedProduct.value.ingredients = []
        ingredients.value = selectedProduct.value.ingredients.map(item => ({
          name: item.name,
          amount: '' + item.amount
        }))
      } else {
        ingredients.value = []
      }
    }

    // reload ingredients each time product changed
    watch(selectedProduct, () => loadIngredients())

    onMounted(async () => {
      /*onCreated*/ loadIngredients();
      await loadCategories()
      await loadProducts()
      await loadInventories()
    })

    const rules = computed(() => {
      let rules = []
      const _inventories = ingredients.value.map(item => {
        const inventory = detailInventories.value.find(invt => invt.product.name === item.name)
        return inventory ? inventory.product.name : ''
      })
      rules.push(val => _inventories.filter(item => item.toString() === val.toString()).length <= 1 || '')
      return rules
    })

    const addIngredient = () => ingredients.value.unshift({ inventory: '', amount: '' })
    const getTouchHandler = (index) => ({
      right: () => ingredients.value.splice(index, 1)
    })

    async function  updateProductIngredient() {
      const _ingredients = ingredients.value.map(item => ({
        inventory: detailInventories.value.find(inventory => inventory.product.name === item.name)._id,
        name: item.name,
        amount: Number(item.amount)
      }))
      if(_.some(_.countBy(_ingredients, 'inventory'), item => item > 1))
        return
      await cms.getModel('Product').findOneAndUpdate({ _id: selectedProduct.value._id }, { ingredients: _ingredients })
      await loadOrderLayout()
    }

    const debounceUpdateAmount = _.debounce(updateProductIngredient, 300)

    //fixme: i18n
    const renderNoProductSelected = () => <div style="display: flex; align-items: center; justify-content: center; height: 100%">
      {t('restaurant.menuEdit.mustSelectProduct')}
    </div>

    const renderIngredientEditor = () => {
      return (
          <div class="ingredient-editor">
            <g-btn-bs style="margin: 0" block background-color="#1271FF" icon="add@16" onClick={addIngredient}>{t('inventory.ingredient')}</g-btn-bs>
            {ingredients.value.map((ingredient, i) =>
                <div v-touch="getTouchHandler(i)" class="ingredient-editor__input" key={i}>
                  <g-autocomplete
                      class="ingredient-editor__input--left"
                      text-field-component="GTextFieldBs"
                      virtual-event
                      menu-class="menu-select-inventory"
                      key={`auto_${ingredients.length - i}`}
                      rules={rules.value}
                      items={productInventoryName.value}
                      arrow={false}
                      v-model={ingredient.name}
                      onInputClick={() => showKeyboard.value = true}
                      onUpdate:modelValue={updateProductIngredient}/>
                  <g-text-field-bs
                      rules={[val => !isNaN(val) || '']}
                      class="ingredient-editor__input--right"
                      virtual-event
                      v-model={ingredient.amount}
                      onClick={() => showKeyboard.value = true}
                      onUpdate:modelValue={debounceUpdateAmount}/>
                </div>
            )}
            <div class="ingredient-editor__message">
              {t('inventory.swipeRight')}
              <g-icon style="margin-bottom: 2px" color="#757575" size="16">fas fa-angle-double-right</g-icon>
            </div>
            {
              (showKeyboard.value) && <>
                <div class="ingredient-editor__keyboard">
                  <div class="ingredient-editor__overlay" onClick={() => showKeyboard.value = false}></div>
                  <div class="ingredient-editor__keyboard-wrapper">
                    <pos-keyboard-full onEnterPressed={() => showKeyboard.value = false}></pos-keyboard-full>
                  </div>
                </div>
              </>
            }
          </div>
      )
    }

    return genScopeId(() => selectedProductExisted.value ? renderIngredientEditor() : renderNoProductSelected())
  }
}
</script>
<style scoped lang="scss">
.ingredient-editor {
  padding: 8px;

  &__input {
    display: flex;
    margin-top: 8px;

    &--left, &--right {
      margin: 0;
      background-color: #F0F0F0;

      ::v-deep .bs-tf-input {
        background: transparent;
      }
    }

    &--left {
      flex: 0 0 70%;

      ::v-deep .bs-tf-wrapper {
        margin: 0;
        width: 100%;

        .bs-tf-inner-input-group {
          border-radius: 4px 0 0 4px;
          border-right: none;
        }
      }
    }

    &--right {
      flex: 0 0 30%;

      ::v-deep .bs-tf-inner-input-group {
        border-radius: 0 4px 4px 0;
      }
    }
  }

  &__message {
    margin-top: 4px;
    font-size: 13px;
    font-weight: 700;
    font-style: italic;
    color: #757575;
  }

  &__keyboard {
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 30%;
    z-index: 10;
    display: flex;
    flex-direction: column;

    &-wrapper {
      padding: 6px;
      background: #bdbdbd;
    }
  }

  &__overlay {
    flex: 1;
    background: rgba(107, 111, 130, 0.7);
  }
}
</style>
