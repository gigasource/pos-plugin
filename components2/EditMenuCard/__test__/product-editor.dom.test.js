/**
 * NOTE:
 * This file contain test cases to test whether ProductEditor render correct or not
 *
 * With valid product layout, it should render product basic info section in the top to allow
 * the user to modify the (name, id, price, isModifier, ...)
 *
 * It's also render printer section which:
 *    - display list of printers,
 *    - a +2 printer button to allow the user
 *    - a no printer button to select none of them
 *
 * It's also display taxes for dineIn tax and takeAway tax
 * It's also display pop-up modifier section to allow the custom or selected predefined modifier
 *
 * -----------
 *
 * It's doesn't contain any test cases to test actions (delete, switch, copy) or change product information
 * It's assume that dependencies component (portal, 'dialog-confirm-delete', 'g-grid-select',
 *    'color-selector','input-number','dialog-text-filter', ... are render correctly with correct input.
 *
 * Issue:
 *   Even though the printers has been set, snapshot doesn't display list of printers (-_- )
 */

import { wrapper, makeWrapper } from "../../test-utils";
import { mockProduct } from "../../OrderView/__test__/mock_product";
const delay = require("delay");
import _ from 'lodash'
import ProductEditor2 from '../ProductEditor/ProductEditor2';
import {nextTick} from 'vue'
import {
  orderLayout, selectCategoryLayout, selectProductLayout,
} from '../../OrderView/pos-ui-shared';
import StubFactory from '../../StubFactory';
import { createEmptyLayout, createEmptyProductLayout } from '../utils';
import { allowSelectPrinter2, dineInTaxes, printers, takeAwayTaxes } from '../ProductEditor/ProductEditorLogic';
const ignoreComponents = ['Portal', 'PortalTarget', 'MountingPortal', 'portal', 'portal-target']

// mock taxes
const mockDineInTaxes = [{ _id: 'di19', item: '19%', value: 19 }, { _id: 'di7', item: '7%', value: 7 }]
const mockTakeAwayTaxes = [{  _id: 'ta19', item: '19%', value: 19 }, { _id: 'ta7', item: '7%', value: 7 }]

// mock printer
const mockPrinters = [{_id: 'pr1', name: 'bar'}, { _id: 'pr2', name: 'kitchen' }]

// mock modifier

describe('product-editor', () => {
  function setupProduct(product, productLayout = {}) {
    // setup order layout
    orderLayout.value = { rows: 1, columns: 3, type: 'default', categories: [
        {
          row: 1, column: 2, top: 0, left: 0,
          products: [
            { product, top: 0, left: 0, ...productLayout}
          ]
        }
      ]}
    selectCategoryLayout({ top: 0, left: 0 })
    selectProductLayout({ top: 0, left: 0 })
  }

  async function expectSnapshot() {
    // wipe-out components (to make dependencies component un-resolvable) to make sure snapshot render prop for stub class
    ProductEditor2.components = {}
    makeWrapper(ProductEditor2, {
      shallow: false,
      global: {
        renderStubDefaultSlot: false,
        stubs: {
          portal: true,
          'dialog-confirm-delete': true,
          'color-selector': true,
          'input-number': true,
          'dialog-text-filter': true,
          'g-select': true,
          'g-switch': true,
          'dialog-product-info': true,
          'dialog-edit-popup-modifiers': true
        },
        config: {
          isCustomElement: tag => _.includes(ignoreComponents, tag)
        }
      }
    }, false)

    await nextTick()
    await delay(50)
    expect(wrapper.html()).toMatchSnapshot()
  }

  async function setupEntireRender(product) {
    setupProduct(product)
    await expectSnapshot()
  }

  it('should render text', async () => {
    setupProduct(null, { type: 'Text', name: 'Soft Drink' })
    await expectSnapshot()
  })

  it('should render placeholder product layout', async () => {
    await setupEntireRender({...createEmptyLayout(), ...createEmptyProductLayout()})
  })

  it('should render real product', async () => {
    await setupEntireRender(mockProduct)
  })

  describe('[printer]', () => {
    // entire printer setting won't be render
    it('should not render printer section if product is modifier', async () => {
      await setupEntireRender({ ...mockProduct, isModifier: true })
    })

    describe('[Add printer 2 btn]', () => {
      // printers.value = mockPrinters

      it('should show when product only have 1 printer', async () => {
        await setupEntireRender({ ...mockProduct, isModifier: false,
          /* option below indicate that product has only 1 printer */
          isNoPrint: false, groupPrinter: 'bar', groupPrinter2: null
        })
      })

      it('should not show if no printer selected or 2 printer selected', async () => {
        await setupEntireRender({ ...mockProduct, isModifier: false, isNoPrint: true })
        await setupEntireRender({ ...mockProduct, isModifier: false, isNoPrint: false, groupPrinter: 'bar', groupPrinter2: 'kitchen' })
      })

      it('should not show if this button has been clicked', async () => {
        allowSelectPrinter2()
        await setupEntireRender({ ...mockProduct, isModifier: false, isNoPrint: false, groupPrinter: 'bar', groupPrinter2: null })
      })
    })
  })
})
