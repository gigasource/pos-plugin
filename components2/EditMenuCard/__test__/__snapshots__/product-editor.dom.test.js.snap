// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product-editor [printer] [Add printer 2 btn] should not show if no printer selected or 2 printer selected 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="true"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub model-value="48">
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="Whiskey">
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="40">
    </g-text-field-bs-stub>
    <g-switch-stub model-value="false">
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div>
    <div class="product-editor__prop">
      <span class="product-editor__label">
        restaurant.product.printer
      </span>
    </div>
    <div>
      <span class="prop-option prop-option--1">
        restaurant.product.noPrinter
      </span>
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor [printer] [Add printer 2 btn] should not show if no printer selected or 2 printer selected 2`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="true"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub model-value="48">
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="Whiskey">
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="40">
    </g-text-field-bs-stub>
    <g-switch-stub model-value="false">
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div>
    <div class="product-editor__prop">
      <span class="product-editor__label">
        restaurant.product.printer
      </span>
    </div>
    <div>
      <span class="prop-option">
        restaurant.product.noPrinter
      </span>
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor [printer] [Add printer 2 btn] should not show if this button has been clicked 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="true"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub model-value="48">
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="Whiskey">
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="40">
    </g-text-field-bs-stub>
    <g-switch-stub model-value="false">
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div>
    <div class="product-editor__prop">
      <span class="product-editor__label">
        restaurant.product.printer
      </span>
    </div>
    <div>
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor [printer] [Add printer 2 btn] should show when product only have 1 printer 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="true"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub model-value="48">
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="Whiskey">
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="40">
    </g-text-field-bs-stub>
    <g-switch-stub model-value="false">
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div>
    <div class="product-editor__prop">
      <span class="product-editor__label">
        restaurant.product.printer
      </span>
      <span class="prop-option--printer">
        +2. restaurant.product.printer
      </span>
    </div>
    <div>
      <span class="prop-option">
        restaurant.product.noPrinter
      </span>
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor [printer] should not render printer section if product is modifier 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="true"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub model-value="48">
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="Whiskey">
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="40">
    </g-text-field-bs-stub>
    <g-switch-stub model-value="true">
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor should render placeholder product layout 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="false"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub>
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value>
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub>
    </g-text-field-bs-stub>
    <g-switch-stub>
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div>
    <div class="product-editor__prop">
      <span class="product-editor__label">
        restaurant.product.printer
      </span>
    </div>
    <div>
      <span class="prop-option">
        restaurant.product.noPrinter
      </span>
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor should render real product 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="true"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.id
    </div>
    <g-text-field-bs-stub model-value="48">
    </g-text-field-bs-stub>
    <div>
      article.name
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="Whiskey">
    </g-text-field-bs-stub>
    <div>
      article.price
      <span style="color: rgb(255, 68, 82);">
        *
      </span>
    </div>
    <g-text-field-bs-stub model-value="40">
    </g-text-field-bs-stub>
    <g-switch-stub>
    </g-switch-stub>
    <div style="font-size: 13px;">
      article.isModifier
    </div>
  </div>
  <div>
    <div class="product-editor__prop">
      <span class="product-editor__label">
        restaurant.product.printer
      </span>
      <span class="prop-option--printer">
        +2. restaurant.product.printer
      </span>
    </div>
    <div>
      <span class="prop-option">
        restaurant.product.noPrinter
      </span>
    </div>
  </div>
  <div class="row-flex mt-2 product-editor__tax">
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.dineInTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
    <div class="col-6">
      <div class="product-editor__label">
        restaurant.product.takeAwayTax
      </div>
      <g-grid-select-stub mandatory="true"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <div class="mt-2">
    <div class="product-editor__label">
      ui.color
    </div>
    <color-selector-stub colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                         item-size="25"
    >
    </color-selector-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
  <dialog-product-info-stub modelvalue="false"
                            product="[object Object]"
  >
  </dialog-product-info-stub>
  <dialog-text-filter-stub label="Text"
                           modelvalue="false"
  >
  </dialog-text-filter-stub>
  <dialog-edit-popup-modifiers-stub modelvalue="false"
                                    product="[object Object]"
  >
  </dialog-edit-popup-modifiers-stub>
</div>
`;

exports[`product-editor should render text 1`] = `
<div class="product-editor">
  <div class="product-editor__prop-grid">
    <div>
      article.type
    </div>
    <g-select-stub disabled="false"
                   skip-search="true"
                   text-field-component="GTextFieldBs"
                   modelvalue="Text"
                   items="[object Object],[object Object],[object Object]"
    >
    </g-select-stub>
    <div>
      article.name
      <span style="color: rgb(255, 69, 82);">
        *
      </span>
    </div>
    <g-text-field-stub>
    </g-text-field-stub>
  </div>
  <div class="mt-2">
    <div class="row-flex justify-between">
      <div class="product-editor__label">
        Popup modifiers
      </div>
      <g-icon-stub size="16">
        icon-edit_modifiers
      </g-icon-stub>
    </div>
    <div>
      <g-grid-select-stub item-text="name"
                          item-value="_id"
                          items
                          itemcols="auto"
      >
      </g-grid-select-stub>
    </div>
  </div>
</div>
`;
